lib/ColorThief/ColorThief.php

```diff
@@ -137,118 +139,234 @@ public static function getPalette($sourceImage, $colorCount = 10, $quality = 10,
     }
 
     /**
      * @param mixed      $sourceImage Path/URL to the image, GD resource, Imagick instance, or image as binary string
-     * @param int        $quality
+     * @param int        $quality Analyze every $quality pixels
      * @param array|null $area
+     * @param array      $histo Histogram
      *
-     * @return SplFixedArray
+     * @return int
      */
-    private static function loadImage($sourceImage, $quality, array $area = null)
+    private static function loadImage_GD($sourceImage, $quality, $area, array &$histo)
     {
-        $loader = new ImageLoader();
-        $image = $loader->load($sourceImage);
-        $startX = 0;
-        $startY = 0;
-        $width = $image->getWidth();
-        $height = $image->getHeight();
-
-        if ($area) {
-            $startX = isset($area['x']) ? $area['x'] : 0;
-            $startY = isset($area['y']) ? $area['y'] : 0;
-            $width = isset($area['w']) ? $area['w'] : ($width - $startX);
-            $height = isset($area['h']) ? $area['h'] : ($height - $startY);
-
-            if ((($startX + $width) > $image->getWidth()) || (($startY + $height) > $image->getHeight())) {
-                throw new \InvalidArgumentException('Area is out of image bounds.');
+        if ((is_resource($sourceImage) && get_resource_type($sourceImage) == 'gd')) {
+            $image = $sourceImage;
+        } else {
+            // Tries to detect if the source string is a binary string or a path to an existing file
+            // This test is based on the way that PHP detects an invalid path and throws a warning
+            // saying "xxx expects to be a valid path, string given" (see zend_parse_arg_path_str).
+            if (strpos($sourceImage, "\0") !== false) {
+                // Binary string
+                $image = @imagecreatefromstring($sourceImage);
+                if ($image === false) {
+                    throw new \InvalidArgumentException('Passed binary string is empty or is not a valid image');
+                }
+            } else {
+                // Path or URL
+                $is_remote = filter_var($sourceImage, FILTER_VALIDATE_URL);
+                if (!$is_remote && (!file_exists($sourceImage) || !is_readable($sourceImage))) {
+                    throw new \RuntimeException("Image '" . $sourceImage . "' is not readable or does not exists.");
+                }
+
+                list(, , $type) = getimagesize($sourceImage);
+                switch ($type) {
+                    case IMAGETYPE_GIF:
+                        $image = @imagecreatefromgif($sourceImage);
+                        break;
+
+                    case IMAGETYPE_JPEG:
+                        $image = @imagecreatefromjpeg($sourceImage);
+                        break;
+
+                    case IMAGETYPE_PNG:
+                        $image = @imagecreatefrompng($sourceImage);
+                        break;
+
+                    case IMAGETYPE_WEBP:
```

Comment:
Interresting, but please open a separate PR if you want to add support for WEBP. This should be done for each loader if possible, and may need some doc update.